src/EventLoop.hs

```diff
@@ -122,59 +122,57 @@ runGithubEventLoop ghQueue enqueueEvent = runLoop
       runLoop
 
 runLogicEventLoop
-  :: MonadIO m
-  => MonadLogger m
-  => MonadMonitor m
+  ::  forall es
+  .  IOE :> es
+  => Metrics.MetricsOperation :> es
+  => Time.TimeOperation :> es
+  => Git.GitOperation :> es
+  => GithubApi.GithubOperation :> es
+  => MonadLoggerEffect :> es
   => TriggerConfiguration
   -> ProjectConfiguration
   -> MergeWindowExemptionConfiguration
-  -- Interpreters for Git and GitHub actions.
-  -> (forall a. Metrics.MetricsOperationFree a -> m a)
-  -> (forall a. Time.TimeOperationFree a -> m a)
-  -> (forall a. Git.GitOperationFree a -> m a)
-  -> (forall a. GithubApi.GithubOperationFree a -> m a)
   -- Action that gets the next event from the queue.
-  -> m (Maybe Logic.Event)
+  -> IO (Maybe Logic.Event)
   -- Action to perform after the state has changed, such as
   -- persisting the new state, and making it available to the
   -- webinterface.
-  -> (ProjectState -> m ())
+  -> (ProjectState -> IO ())
   -> ProjectState
-  -> m ProjectState
+  -> Eff es ProjectState
 runLogicEventLoop
   triggerConfig projectConfig mergeWindowExemptionConfig
-  runMetrics runTime runGit runGithub
   getNextEvent publish initialState =
   let
     repo             = Config.repository projectConfig
-    runAll           = foldFree (runSum (runSum runMetrics runTime) (runSum runGit runGithub))
-    runAction        = foldFree (runSum (Logic.runBaseAction projectConfig) (Logic.runRetrieveEnvironment projectConfig))
     showState state' = fromRight (showText state') $ decodeUtf8' $ toStrict $ encode state'
     handleAndContinue state0 event = do
       -- Handle the event and then perform any additional required actions until
       -- the state reaches a fixed point (when there are no further actions to
       -- perform).
       logInfoN  $ "logic loop received event (" <> repo <> "): " <> showText event
       logDebugN $ "state before (" <> repo <> "): " <> showState state0
-      state1 <- runAll $ runAction $
+      state1 <-
         Logic.handleEvent triggerConfig mergeWindowExemptionConfig event state0
-      publish state1
+      liftIO $ publish state1
       logDebugN $ "state after (" <> repo <> "): " <> showState state1
       runLoop state1
 
     runLoop state = do
       -- Before anything, clone the repository if there is no clone.
-      foldFree runGit $ Logic.ensureCloned projectConfig
```

Comment:
In order to run git free-monad things, you needed to call them in `runGit`. With `effectful`, the functions can just be called.

Reply:
_Incroyable_.